### PR TITLE
Exclude prefixes or folders from versioning

### PR DESCRIPTION
**Pull request by feorlen**
Object versioning can now be configured with certain exclusions (no longer strictly on a per-bucket basis). Add documentation about how to use the new options:
* `--excluded-prefixes` to exclude objects with certain prefixes
* `--exclude-folders` to exclude anything with name ending in `/` (folders)

- [x] Exclude prefixes
- [x] Exclude folders
- [x] Wildcards/globbing
- [x] Limitations
- [x] View versioning exclusion configuration

Staged:
[Bucket Versioning](http://192.241.195.202:9000/staging/DOCS-760/linux/html/administration/object-management/object-versioning.html)
[mc version](http://192.241.195.202:9000/staging/DOCS-760/linux/html/reference/minio-mc/mc-version.html)

Fixes https://github.com/minio/docs/issues/760

### CODE REVIEW

**Comment by feorlen** [comment]
@ravindk89 @djwfyi First draft for the versioning exclusion docs, very WIP. If you have a moment, comments on usual style and formatting conventions much appreciated.

File: source/administration/object-management/object-versioning.rst [hunk 1]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+    be excluded from versioning.
+
+Something about wildcards and globbing
+    
+You can exclude up to 10 prefixes for each bucket. To add or remove excluded prefixes,
+repeat your ``mc version enable`` command with an updated list. The new list of
+prefixes replaces the previous one.
+
+To view the currently excluded prefixes, use the :mc-cmd:`mc version enable`
+command with the ``--json`` option. The ``ExcludedPrefixes`` property contains a list
```
Comment (djwfyi) [inline source/administration/object-management/object-versioning.rst:299]: You can simplify this, I think to
> To view the excluded prefixes, use :mc-cmd:`mc version enable` with the  :mc-cmd:`~mc  version enable --json` option.

When we use `~` at the start of the :mc-cmd:, it only displays the last element of the command, so we can link directly to the flag in the docs.

File: source/administration/object-management/object-versioning.rst [hunk 2]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+of excluded prefixes:
+
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --excluded-prefixes ALIAS/BUCKET --json
+
+  .. code-block:: shell
+
```
Comment (djwfyi) [inline source/administration/object-management/object-versioning.rst:308]: Add a sentence at line 306 along the lines of
> The output resembles the following:

File: source/administration/object-management/object-versioning.rst [hunk 3]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+        "TempPrefix-"
+       ]
+      }
+     }
+
+
+Exclude Folders from Versioning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can exclude folders from versioning using the MinIO :mc:`mc` CLI.
```
[the remark this answers is not in the document]
Comment (djwfyi) [reply source/administration/object-management/object-versioning.rst:327]: Only, perhaps also spell `versioning` correctly. 🙄

File: source/administration/object-management/object-versioning.rst [hunk 4]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+
+- Use the :mc-cmd:`mc version enable` command with the ``--exclude-folders``
```
Comment (djwfyi) [inline source/administration/object-management/object-versioning.rst:329]: As above, you can provide a link to the `--exclude-folder` option in the docs:
```
:mc-cmd:`~mc version enable --exclude-folders`
```

And I would drop "the" and "command" and just let the command stand on its own.
> Use `mc version enable` with the `--exclude-folders`...

File: source/administration/object-management/object-versioning.rst [hunk 5]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --exclude-folders ALIAS/BUCKET
+
+  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured
+    MinIO deployment.
+
+  - Replace ``BUCKET`` with the
+    :mc-cmd:`target bucket <mc version ALIAS>` for which the folders should
```
Comment (djwfyi) [inline source/administration/object-management/object-versioning.rst:341]: Different link for the bucket here, as well.

File: source/administration/object-management/object-versioning.rst [hunk 6]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+    be excluded from versioning.
+
+To check whether folders are versioned for a bucket, use the :mc-cmd:`mc version enable`
+command with the ``--json`` option. If the ``ExcludeFolders`` property is ``true``,
```
Comment (djwfyi) [inline source/administration/object-management/object-versioning.rst:345]: Similar wording adjustment.

File: source/administration/object-management/object-versioning.rst [hunk 7]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+folders in that bucket are not versioned.
+
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --excluded-prefixes ALIAS/BUCKET --json
+
+  .. code-block:: shell
```
Comment (djwfyi) [inline source/administration/object-management/object-versioning.rst:353]: Add a sentence above intro-ing the example, as in the previous comment above.

**Review by djwfyi — CHANGES_REQUESTED** [review]
Nice work on this PR so far, Andrea!

For general style things we do:
- We can use a `~` with an mc command to only refer to the last element in the command while also generating a link to that command or flag.
- We have moved in the last year to using ventilated prose, one sentence per line.
- Try to avoid passives whenever possible.
- A command itself can be an actor, so we don't need to call it "the `mc alias set` command". Just "`mc alias set`" is fine.
- Provide some kind of intro statement before a sample output.

File: source/administration/object-management/object-versioning.rst (hunk 1 above)
Comment (djwfyi) [reply source/administration/object-management/object-versioning.rst:299]: Except this doesn't work, because --json is a global flag. So ignore this.

File: source/administration/object-management/object-versioning.rst [hunk 8]
```diff
@@ -266,6 +266,105 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+     :class: copyable
+
+     mc version enable --excluded-prefixes ALIAS/BUCKET "prefix1, prefix2"
+
+  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured 
+    MinIO deployment.
+
+  - Replace ``BUCKET`` with the 
+    :mc-cmd:`target bucket <mc version ALIAS>` for which the prefixes should
+    be excluded from versioning.
```
[the remark this answers is not in the document]
Comment (feorlen) [reply source/administration/object-management/object-versioning.rst:290]: The existing text just above has the same 
```
:mc-cmd:`target bucket <mc version ALIAS>`
```
maybe that should change too

File: source/reference/minio-mc/mc-version.rst [hunk 9]
```diff
@@ -60,6 +60,11 @@ Parameters
    Mutually exclusive with :mc-cmd:`~mc version suspend` and
    :mc-cmd:`~mc version info`
 
+   .. mc-cmd:: enable --excluded-prefixes
```
Comment (ravindk89) [inline source/reference/minio-mc/mc-version.rst:63]:
```suggestion
.. mc-cmd:: --excluded-prefixes
```

Since this is nested, it inherits the parent - going all the way up the chain if necessary

File: source/administration/object-management/object-versioning.rst [hunk 10]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option:
+
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --excluded-prefixes ALIAS/BUCKET "prefix1, prefix2"
+
+  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured MinIO deployment.
+
+  - Replace ``BUCKET`` with the :mc-cmd:`target bucket <mc version ALIAS>` you want to exclude :ref:`prefixes <minio-admin-concepts-organize-objects>` for.
```
Comment (feorlen) [inline source/administration/object-management/object-versioning.rst:289]: I tried to both follow the model of the existing example above, and link to something useful about prefixes/folders. Ideas? This seems a bit awkward.

File: source/administration/object-management/object-versioning.rst [hunk 11]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+
+The list of :mc-cmd:`~mc version --excluded-prefixes` prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``*prefix*``.
+To match objects by prefix only, use ``prefix/*``.
+
+For example, the following command excludes any objects containing ``abc`` or ``d3f`` in their prefix or name from versioning:
+
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable local/test-bucket --excluded-prefixes "abc, d3f"
```
Comment (feorlen) [inline source/administration/object-management/object-versioning.rst:299]: Is there an established pattern for names/tokens in examples? (Rather than abc, etc.)

File: source/administration/object-management/object-versioning.rst [hunk 12]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+
+The command output resembles the following:
+
+.. code-block:: shell
+
+     $ mc version info local/test-bucket --json
+     {
+      "Op": "info",
+      "status": "success",
+      "url": "local/test-bucket",
```
Comment (feorlen) [inline source/administration/object-management/object-versioning.rst:375]: Is there a standard example bucket name I might use here?

File: source/reference/minio-mc/mc-version.rst [hunk 13]
```diff
@@ -60,6 +60,18 @@ Parameters
    Mutually exclusive with :mc-cmd:`~mc version suspend` and
    :mc-cmd:`~mc version info`
 
+.. mc-cmd:: --exclude-folders
```
Comment (feorlen) [inline source/reference/minio-mc/mc-version.rst:63]: Is there a usual way to order parameters in the reference?

I put these two directly under `enable` because they are children, despite it not being obvious.

Examples would help (oops forgot to add) but is there more generally a better way? 

Besides reworking the entire page to get the parent/child relationships correct, as we discussed in slack.

File: source/reference/minio-mc/mc-version.rst [hunk 14]
```diff
@@ -78,7 +90,9 @@ Parameters
 
 .. mc-cmd:: ALIAS
 
-   *Required* The :ref:`alias <alias>` of a MinIO deployment and the full path
+   *Required*
```
Comment (feorlen) [inline source/reference/minio-mc/mc-version.rst:93]: A little drive-by formatting to match the other items in this section.

`:required:` has a very different style, maybe that could be updated. (If it doesn't globally upset too many things.)

**Comment by feorlen** [comment]
Note for reviewers:

Now this PR is at a reasonable state to do a proper review. Added comments about a few things I had questions about. Not yet staged, need to try that on Monday.

File: source/administration/object-management/object-versioning.rst (hunk 10 above)
Comment (djwfyi) [reply source/administration/object-management/object-versioning.rst:289]: Couple of options here.

## Link to the S3 Docs
`Bucket` is a standard s3 concept. We could link to their docs explanation.
If you look in conf.py under extlinks, we have a role for that:
```
:s3-docs:`UsingBucket.html`
```

Basically, s3-docs adds the base URL up through `/latest/userguide/` then attaches whatever we have for the rest of the URL.

## Link to the list command

We could link to the mc list command:
```
:mc:`mc ls`
```

(That would list all buckets and objects at the root of the alias.)

## Link to the admin/concept.rst page

There's a section on that page about organizing objects within folders in buckets. Not ideal, but closest we have there. You'd need to add a directive to that page."

## My preference
I think I prefer linking to the S3 docs, personally.

File: source/administration/object-management/object-versioning.rst (hunk 11 above)
Comment (djwfyi) [reply source/administration/object-management/object-versioning.rst:299]: `prefix1, prefix2`
or
`myprefix, mysecondprefix`

Something that is human readable.

We typically use the word for whatever it is we are samplig. `mybucket`, `myminio` (for an alias), and so on.

File: source/administration/object-management/object-versioning.rst (hunk 12 above)
Comment (djwfyi) [reply source/administration/object-management/object-versioning.rst:375]: Most often, we just go with `my-bucket`.

File: source/reference/minio-mc/mc-version.rst (hunk 13 above)
Comment (djwfyi) [reply source/reference/minio-mc/mc-version.rst:63]: Parameters are supposed to go:
- Required params, alphabetically
   With the `:required:` class.
- Optional params, alphabetically
   With the `:optional:` class

I think we are still pretty inconsistent with this at the moment, but that's the order we're trying to get to.

File: source/reference/minio-mc/mc-version.rst (hunk 14 above)
Comment (djwfyi) [reply source/reference/minio-mc/mc-version.rst:93]: Drive by fixes you note to do are always welcome!

I say it's better to go with `:required:` here and let's get the css fixed. Better than having to go back and do it later. We can live with a momentary problematic formatting until the CSS fixes go in.

@rushenn for your eyes.

Comment (feorlen) [reply source/reference/minio-mc/mc-version.rst:93]: The existing parameters on this page don't have `:optional:` or `:required:` (only the one manually styled "required") so this would be a good time to deal with all of them together.

**Comment by feorlen** [comment]
Staged:
[Bucket Versioning](http://192.241.195.202:9000/staging/DOCS-760/linux/html/administration/object-management/object-versioning.html)
[mc version](http://192.241.195.202:9000/staging/DOCS-760/linux/html/reference/minio-mc/mc-version.html)

File: source/administration/object-management/object-versioning.rst [hunk 15]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can exclude folders from versioning using the :ref:`MinIO Client <minio-client>`.
+
+.. admonition:: Object locking
+   :class: note
+
+   Buckets with :ref:`object locking enabled <minio-object-locking>` require versioning and do not support excluding folders.
+
+- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option to exclude objects with names ending in ``/`` from versioning:
```
Comment (ravindk89) [inline source/administration/object-management/object-versioning.rst:348]: We can use a more 'real world' example here.

e.g.

The following :mc-cmd:`mc version enable` command includes the :mc-cmd:`~mc verison --excluded-prefixes` option to exclude objects in the ``_temp`` prefix from versioning:

```
mc version enable --exclude-folders "_temp" myminio/mybucket
```

Or something along those lines. We get to be creative with examples, so feel free to have a bit of fun thinking of an interesting scenario to build the example around.

File: source/administration/object-management/object-versioning.rst [hunk 16]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+   :class: note
+
+   Buckets with :ref:`object locking enabled <minio-object-locking>` require versioning and do not support excluding prefixes.
+
+- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option:
+
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --excluded-prefixes ALIAS/BUCKET "prefix1, prefix2"
```
Comment (ravindk89) [inline source/administration/object-management/object-versioning.rst:285]: I think the syntax has to be:

```suggestion
mc version enable --excluded-prefixes "prefix1,prefix2" ALIAS/BUCKET
```

```
mc version enable --excluded-prefixes play/version-test "foo"                             1 ✘  docs   3.8.16   
mc: <ERROR> Unable to enable versioning: `SetVersion` is not supported for `filesystem`.
```

I _think_ we generally recommend doing all params prior to the ALIAS/BUCKET for parsing reasons, but I am not entirely sure if that's stlil a requirement.

File: source/administration/object-management/object-versioning.rst (hunk 12 above)
Comment (ravindk89) [reply source/administration/object-management/object-versioning.rst:375]: my-bucket, data, etc.

**Review by ravindk89 — CHANGES_REQUESTED** [review]
A few things to tighten - and a general suggestion to feel comfortable with creating interesting examples, since it lets us flex a bit of creative muscle.

File: source/reference/minio-mc/mc-version.rst (hunk 13 above)
Comment (feorlen) [reply source/reference/minio-mc/mc-version.rst:74]: Sorted alpha within _required_ or _optional_.`--` is ignored.

File: source/reference/minio-mc/mc-version.rst (hunk 14 above)
Comment (feorlen) [reply source/reference/minio-mc/mc-version.rst]: Current code in repo uses the first, italic and following text below.

**Review by ravindk89 — APPROVED** [review]

File: source/administration/object-management/object-versioning.rst [hunk 17]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --excluded-prefixes "prefix1, prefix2" ALIAS/BUCKET
+
+  - Replace ``ALIAS`` with the :mc:`alias <mc alias>` of a configured MinIO deployment.
+
+  - Replace ``BUCKET`` with the name of the :s3-docs:`bucket <UsingBucket.html>` you want to exclude :ref:`prefixes <minio-admin-concepts-organize-objects>` for.
+
+The list of :mc-cmd:`~mc version --excluded-prefixes` prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``*prefix*``.
```
Comment (krisis) [inline source/administration/object-management/object-versioning.rst:291]: In this example, the given excluded prefixes `prefix1`, `prefix2` would match objects similar to the regular expression of the form `prefix*`.
```suggestion
The list of :mc-cmd:`~mc version --excluded-prefixes` prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``prefix*``.
```

File: source/administration/object-management/object-versioning.rst [hunk 18]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+  .. code-block:: shell
+     :class: copyable
+
+     mc version enable --excluded-prefixes "_test, _temp" local/my-bucket
+
+You can exclude up to 10 prefixes for each bucket.
+To add or remove prefixes, repeat the :mc-cmd:`mc version enable` command with an updated list.
+The new list of prefixes replaces the previous one.
+
+To view the currently excluded prefixes, use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version enable --JSON` option:
```
Comment (krisis) [inline source/administration/object-management/object-versioning.rst:305]:
```suggestion
To view the currently excluded prefixes, use :mc-cmd:`mc version info` with the :mc-cmd:`~mc version enable --JSON` option:
```

File: source/administration/object-management/object-versioning.rst [hunk 19]
```diff
@@ -266,6 +265,129 @@ enable versioning on only a prefix or subset of objects in a bucket.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can exclude folders from versioning using the :ref:`MinIO Client <minio-client>`.
+
+.. admonition:: Object locking
+   :class: note
+
+   Buckets with :ref:`object locking enabled <minio-object-locking>` require versioning and do not support excluding folders.
+
+- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --excluded-prefixes` option to exclude objects with names ending in ``/`` from versioning:
```
Comment (krisis) [inline source/administration/object-management/object-versioning.rst:348]:
```suggestion
- Use :mc-cmd:`mc version enable` with the :mc-cmd:`~mc version --exclude-folders` option to exclude objects with names ending in ``/`` from versioning:
```

File: source/reference/minio-mc/mc-version.rst [hunk 20]
```diff
@@ -52,38 +52,61 @@ The :mc:`mc version` command enables, suspends, and retrieves the
...
-   :mc-cmd:`ALIAS <mc version ALIAS>`.
+   Use with :mc-cmd:`mc version enable` to disable versioning on all folders (objects whose name ends with ``/``) in the specified bucket.
 
-   Mutually exclusive with :mc-cmd:`~mc version suspend` and
-   :mc-cmd:`~mc version info`
+.. mc-cmd:: --excluded-prefixes
+   :optional:
+
+   Use with :mc-cmd:`mc version enable` to disable versioning on objects matching a list of prefixes, up to 10.
+   The list of prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``*prefix*``.
```
Comment (krisis) [inline source/reference/minio-mc/mc-version.rst:83]:
```suggestion
   The list of prefixes match all objects containing the specified strings in their prefix or name, similar to a regular expression of the form ``prefix*``.
```

File: source/administration/object-management/object-versioning.rst (hunk 19 above)
Comment (feorlen) [reply source/administration/object-management/object-versioning.rst:348]: @krisis Oooh missed that paste failure thank you!!
